File: old_contents/boj_codes/C++17/1389.cpp
// https://www.acmicpc.net/problem/1389
// First Written : 20201231
// Last Modified : 20201231

#include <cstdio>
#include <vector>

using namespace std;

struct Edge{
    int start;
    int end;
};

class Graph{
private:
    int N_node;
    vector<vector<int>> dist;
public:
    Graph(int N_user){
        dist.resize(N_user + 1);
        for(auto it = dist.begin(); it!=dist.end(); it++){
            (*it).resize(N_user + 1, 200);
        }

        for(int i = 1; i <= N_user; i++){
            dist[i][i] = 0;
        }
            
        this->N_node = N_user;
    }

    void add_edge(Edge e){
        dist[e.start][e.end] = 1;
        dist[e.end][e.start] = 1;
    }

    void floydd(){
        for(int k = 1; k <= N_node; k++){
            for(int i = 1; i <= N_node; i++){
                for(int j = 1; j <= N_node; j++){
                    if(dist[i][j] > dist[i][k] + dist[k][j])
                        dist[i][j] = dist[i][k] + dist[k][j];
                }
            }
        }
    }

    int get_kevin(){
        int min = 20000;
        int result;
        int kevin;

        for(int i = 1; i <= N_node; i++){
            result = 0;
            for(int j = 1; j <= N_node; j++){
                result += dist[i][j];
            }

            if(result < min){
                min = result;
                kevin = i;
                                
            }

        }
    
        return kevin;
    }
};

int main(){
    int N_user, N_edge;
    scanf("%d %d", &N_user, &N_edge);

    Graph graph = Graph(N_user);
    Edge e;

    while(N_edge--){
        scanf("%d %d", &e.start, &e.end);
        graph.add_edge(e);
    }

    graph.floydd();
    printf("%d", graph.get_kevin());

    return 0;
}

File: old_contents/boj_codes/Python3/1389.py
#https://www.acmicpc.net/problem/1389
# 20200910

import sys

N_user, N_rel = map(int, sys.stdin.readline().split())

class Graph:
    def __init__(self, N_user):
        self.N_user = N_user
        self.kevin = [0] * N_user
        self.d = []
        for i in range(N_user):
            temp = [200] * N_user
            temp[i] = 0
            self.d.append(temp)


    def add_relation(self, X, Y):
        self.d[X - 1][Y - 1] = 1
        self.d[Y - 1][X - 1] = 1

    def floydd(self):
        for k in range(self.N_user):
            for i in range(self.N_user):
                for j in range(self.N_user):
                    if self.d[i][j] > self.d[i][k] + self.d[k][j]:
                        self.d[i][j] = self.d[i][k] + self.d[k][j]

    def get_kevin(self):
        for i in range(N_user):
            self.kevin[i] = sum(self.d[i])

        minimum = min(self.kevin)
        return self.kevin.index(minimum) + 1



graph = Graph(N_user)
for _ in range(N_rel):
    X, Y = map(int, sys.stdin.readline().split())
    graph.add_relation(X, Y)

graph.floydd()
print(graph.get_kevin())
